### DISCUSSION TITLE
Linking raylib against ANGLE on MacOS

### DISCUSSION THREAD

**Discussion by grplyler** [raysan5/raylib#2998] [Q&A] [answered]
Hi There! I am trying to compile and link raylib against Google's ANGLE so I can get better performance on Mac M1 Chips. With OpenGL being deprecated and being about 10 times slower than Metal on M1 Macs, the improved support for ANGLE, essentially giving raylib a Metal Backend is exciting!

**System**
- Macbook Pro M1 (Apply M1 Pro)
- MacOS Monterrey 12.5

**Folder Structure (up to 2 levels)**
```shell
λ ~/code/projects/raylib-cmake-starter/ angle* tree -L 2
.
├── CMakeLists.txt
├── README.md
├── bootstrap.sh
├── build
│   ├── CMakeCache.txt
│   ├── CMakeFiles
│   ├── CPackConfig.cmake
│   ├── CPackSourceConfig.cmake
│   ├── Makefile
│   ├── cmake_install.cmake
│   ├── raylib-starter
│   └── vendor
├── build.sh
├── build_with_angle.sh
├── run.sh
├── src
│   └── main.cpp
└── vendor
    ├── angle
    └── raylib
```

(All paths relative to root project directory unless absolute)
**Background**
1. ANGLE compiled successfully at vendor/angle/out/Release. All libs and includes look fine and in the right place.
2. Modified `vendor/raylib/src/Makefile` and set `GRAPHICS = GRAPHICS_API_OPENGL_ES2` (followed the comment to uncomment to enabled ANGLE by using OpenGL ES 2.0)
3. Ran `make` inside `vendor/raylib/src` and raylib seemed to compile successfully.

**CMakeFiles.txt and output**
This is 1 version of my CMakeLists.txt in the root of my project:
```cmake
cmake_minimum_required(VERSION 3.16)

project(raylib-starter)

set(CMAKE_CXX_STANDARD 14)

file(GLOB_RECURSE SOURCES "src/*.cpp")
add_executable(${PROJECT_NAME} ${SOURCES})

# Set raylib and ANGLE include and library paths
set(RAYLIB_INCLUDE_DIR vendor/raylib/src)
set(RAYLIB_LIBRARY_DIR vendor/raylib/src)
set(ANGLE_INCLUDE_DIR vendor/angle/include)
set(ANGLE_LIBRARY_DIR vendor/angle/out/Release)

# Replace <path_to_angle_repository> with the actual path to your cloned ANGLE repository

target_include_directories(${PROJECT_NAME} PRIVATE
    ${RAYLIB_INCLUDE_DIR}
    ${ANGLE_INCLUDE_DIR}
)

# Find the raylib and ANGLE libraries
find_library(RAYLIB_LIBRARY libraylib.a PATHS ${RAYLIB_LIBRARY_DIR})
find_library(ANGLE_GLESv2_LIBRARY libGLESv2.dylib PATHS ${ANGLE_LIBRARY_DIR})
find_library(ANGLE_EGL_LIBRARY libEGL.dylib PATHS ${ANGLE_LIBRARY_DIR})

# Link required system frameworks for macOS
find_library(COCOA_LIBRARY Cocoa)
find_library(OPENGL_LIBRARY OpenGL)
find_library(IOKIT_LIBRARY IOKit)
find_library(CORE_FOUNDATION_LIBRARY CoreFoundation)
find_library(CORE_VIDEO_LIBRARY CoreVideo)

target_link_libraries(${PROJECT_NAME}
    # Link raylib static library
    ${RAYLIB_LIBRARY}
    # Link ANGLE libraries
    ${ANGLE_GLESv2_LIBRARY}
    ${ANGLE_EGL_LIBRARY}
    # Link macOS system frameworks
    ${OPENGL_LIBRARY}
    ${COCOA_LIBRARY}
    ${IOKIT_LIBRARY}
    ${CORE_FOUNDATION_LIBRARY}
    ${CORE_VIDEO_LIBRARY}
)
```

It compiles fine, but raylib gives a black screen:
```
INFO: Initializing raylib 4.2
INFO: Supported raylib modules:
INFO:     > rcore:..... loaded (mandatory)
INFO:     > rlgl:...... loaded (mandatory)
INFO:     > rshapes:... loaded (optional)
INFO:     > rtextures:. loaded (optional)
INFO:     > rtext:..... loaded (optional)
INFO:     > rmodels:... loaded (optional)
INFO:     > raudio:.... loaded (optional)
INFO: DISPLAY: Device initialized successfully
INFO:     > Display size: 1728 x 1117
INFO:     > Screen size:  800 x 450
INFO:     > Render size:  800 x 450
INFO:     > Viewport offsets: 0, 0
INFO: GL: Supported extensions count: 0
INFO: GL: OpenGL device information:
INFO:     > Vendor:   (null)
INFO:     > Renderer: (null)
INFO:     > Version:  (null)
INFO:     > GLSL:     (null)
WARNING: TEXTURE: Failed to load texture
WARNING: TEXTURE: Failed to load default texture
WARNING: SHADER: [ID 0] Failed to compile vertex shader code
WARNING: SHADER: [ID 0] Failed to compile fragment shader code
WARNING: SHADER: [ID 0] Failed to link shader program
WARNING: SHADER: [ID 0] Failed to load default shader
INFO: RLGL: Render batch vertex buffers loaded successfully in RAM (CPU)
INFO: RLGL: Render batch vertex buffers loaded successfully in VRAM (GPU)
INFO: RLGL: Default OpenGL state initialized successfully
WARNING: TEXTURE: Failed to load texture
INFO: FONT: Default font loaded successfully (224 glyphs)
INFO: TIMER: Target time per frame: 16.667 milliseconds
INFO: FILEIO: [/Library/Fonts/SF-Pro.ttf] File loaded successfully
WARNING: TEXTURE: Failed to load texture
INFO: FONT: Data loaded successfully (64 pixel size | 95 glyphs)
```

However in version of my `CMakeLists.txt` which comments out the System OpenGL implementations:

```cmake
cmake_minimum_required(VERSION 3.16)

project(raylib-starter)

set(CMAKE_CXX_STANDARD 14)

file(GLOB_RECURSE SOURCES "src/*.cpp")
add_executable(${PROJECT_NAME} ${SOURCES})

# Set raylib and ANGLE include and library paths
set(RAYLIB_INCLUDE_DIR vendor/raylib/src)
set(RAYLIB_LIBRARY_DIR vendor/raylib/src)
set(ANGLE_INCLUDE_DIR vendor/angle/include)
set(ANGLE_LIBRARY_DIR vendor/angle/out/Release)

# Replace <path_to_angle_repository> with the actual path to your cloned ANGLE repository

target_include_directories(${PROJECT_NAME} PRIVATE
    ${RAYLIB_INCLUDE_DIR}
    ${ANGLE_INCLUDE_DIR}
)

# Find the raylib and ANGLE libraries
find_library(RAYLIB_LIBRARY libraylib.a PATHS ${RAYLIB_LIBRARY_DIR})
find_library(ANGLE_GLESv2_LIBRARY libGLESv2.dylib PATHS ${ANGLE_LIBRARY_DIR})
find_library(ANGLE_EGL_LIBRARY libEGL.dylib PATHS ${ANGLE_LIBRARY_DIR})

# Link required system frameworks for macOS
find_library(COCOA_LIBRARY Cocoa)
# find_library(OPENGL_LIBRARY OpenGL)
find_library(IOKIT_LIBRARY IOKit)
find_library(CORE_FOUNDATION_LIBRARY CoreFoundation)
find_library(CORE_VIDEO_LIBRARY CoreVideo)

target_link_libraries(${PROJECT_NAME}
    # Link raylib static library
    ${RAYLIB_LIBRARY}
    # Link ANGLE libraries
    ${ANGLE_GLESv2_LIBRARY}
    ${ANGLE_EGL_LIBRARY}
    # Link macOS system frameworks
    # ${OPENGL_LIBRARY}
    ${COCOA_LIBRARY}
    ${IOKIT_LIBRARY}
    ${CORE_FOUNDATION_LIBRARY}
    ${CORE_VIDEO_LIBRARY}
)

```

I get the following errors:

```shell
[ 50%] Linking CXX executable raylib-starter
Undefined symbols for architecture arm64:
  "_glClearDepth", referenced from:
      _rlglInit in libraylib.a(rcore.c.o)
  "_glGetTexImage", referenced from:
      _rlReadTexturePixels in libraylib.a(rcore.c.o)
  "_glPolygonMode", referenced from:
      _rlEnableWireMode in libraylib.a(rcore.c.o)
      _rlDisableWireMode in libraylib.a(rcore.c.o)
ld: symbol(s) not found for architecture arm64
```

Would appreciate it if anyone has anyone has any thoughts on this or knows how to get raylib working with ANGLE!
Thanks for your time!

**Comment by raysan5** [comment]
@grplyler Did you manage to get it working? ANGLE support has been added recently and it's still quitte experimental, not sure if anyone has tried it with macOS M1 CPU just yet.

Please, if you investigated it, let me know your progress!

**Reply by grplyler** [reply]
Hi @raysan5, I didn't quite get it working, but I took a break from it. I can take another stab at it in the next week or so. I'd like to get it working because Metal is capable of about 10x more draw calls than OpenGL on modern Macs. I'll let you know! Thanks for your time.

**Reply by raysan5** [reply]
Nice! Thanks for testing! Take your time! Let me know if you get it working! :)

**Reply by grplyler** [reply, marked as answer]
Update: After a few black screens, thanks to raylibs trace logs, I did manage to get it working on a small 2D project for work and went from about ~3000 FPS (optimized build without ANGLE) to ~7300 FPS with ANGLE!

**Without ANGLE (optimized build)**
<img width="699" alt="Pasted Graphic 3" src="https://github.com/raysan5/raylib/assets/3348536/1a4bf951-e1dd-4770-bc28-f2e74e28d4e1">

**With ANGLE (optimized build, dylib, 2021 Apple M1 Pro, 10 CPU Cores, 10 GPU Cores)**
<img width="685" alt="HetMapper" src="https://github.com/raysan5/raylib/assets/3348536/9fce596e-2c6b-4aff-a693-19e08733d8a8">

**Raylib Debug Tracing with ANGLE**

```
INFO: Initializing raylib 4.6-dev
INFO: Supported raylib modules:
INFO:     > rcore:..... loaded (mandatory)
INFO:     > rlgl:...... loaded (mandatory)
INFO:     > rshapes:... loaded (optional)
INFO:     > rtextures:. loaded (optional)
INFO:     > rtext:..... loaded (optional)
INFO:     > rmodels:... loaded (optional)
INFO:     > raudio:.... loaded (optional)
INFO: DISPLAY: Device initialized successfully
INFO:     > Display size: 1920 x 1080
INFO:     > Screen size:  800 x 450
INFO:     > Render size:  800 x 450
INFO:     > Viewport offsets: 0, 0
INFO: GLAD: OpenGL ES2.0 loaded successfully
INFO: GL: Supported extensions count: 104
INFO: GL: OpenGL device information:
INFO:     > Vendor:   Google Inc. (Apple)
INFO:     > Renderer: ANGLE (Apple, Apple M1 Pro, OpenGL 4.1 Metal - 76.3)
INFO:     > Version:  OpenGL ES 3.0.0 (ANGLE 2.1.21199 git hash: b0e9bbd79fb6)
INFO:     > GLSL:     OpenGL ES GLSL ES 3.00 (ANGLE 2.1.21199 git hash: b0e9bbd79fb6)
INFO: GL: VAO extension detected, VAO functions loaded successfully
INFO: GL: NPOT textures extension detected, full NPOT textures supported
INFO: TEXTURE: [ID 1] Texture loaded successfully (1x1 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 1] Default texture loaded successfully
INFO: SHADER: [ID 1] Vertex shader compiled successfully
INFO: SHADER: [ID 2] Fragment shader compiled successfully
INFO: SHADER: [ID 3] Program shader loaded successfully
INFO: SHADER: [ID 3] Default shader loaded successfully
INFO: RLGL: Render batch vertex buffers loaded successfully in RAM (CPU)
INFO: RLGL: Render batch vertex buffers loaded successfully in VRAM (GPU)
INFO: RLGL: Default OpenGL state initialized successfully
INFO: TEXTURE: [ID 2] Texture loaded successfully (128x128 | GRAY_ALPHA | 1 mipmaps)
INFO: FONT: Default font loaded successfully (224 glyphs)
INFO: SYSTEM: Decompress data: Comp. size: 1104 -> Original size: 131072
INFO: TEXTURE: [ID 3] Texture loaded successfully (256x256 | GRAY_ALPHA | 1 mipmaps)
```

I plan on writing a proper guide for now to build and link ANGLE from scratch to raylib along with benchmarks, since this took a fair bit of googling, trial and error. The dynamic linking is convenient since people can just plop `libEGL.dylib` and `libGLESv2.dylib` into the folder with their raylib project binary. 

But these speed boosts on Mac could really make things interesting. For now, this is an easier solution than using something like `sokol_gfx.h` (which sits at the rlGL) layer. Though I might hack around later with that because i want try Metal's Ray Tracing Hardware built into their modern chips.

I'll be back with a writeup/howto, CMakeLists.txt, and maybe update my raylib starter project when I get some more time!

Thanks!

**Reply by raysan5** [reply]
WOW! This is really AMAZING! Congratulations! IT looks fantastic! A writeup or a simple guide or an update on raylib Wiki would be really appreciated! It could be REALLY useful for many macOS developers! Actually, I'm releasing my tools for macOS very soon and ANGLE could keep them alive if OpenGL gets removed at some point! Thanks again for such great news! 😄

**Comment by chriscamacho** [comment]
last time I looked at Angle on Linux I failed to even get it to compile, so well done, I look forward to the write up so I can have another go!

**Reply by Peter0x44** [reply]
I gave up building ANGLE myself and simply used the so/dll from an installed browser.
It really seems like an enormous pain to do so.

/usr/lib/chromium/libEGL.so
/usr/lib/chromium/libGLESv2.so

I suggest sparing yourself that pain and just linking to these directly.

I also did this on windows, copying the dlls out from edge. They were located in:
C:/Program Files (x86)/Microsoft/Edge/Application/112.0.1722.48/libEGL.dll
C:/Program Files (x86)/Microsoft/Edge/Application/112.0.1722.48/libGLESv2.dll

I got it working by compiling raylib with GRAPHICS=GRAPHICS_API_OPENGL_ES2, and then building the program like this:

`gcc whatever.c -I raylib/src libEGL.dll libGLESv2.dll raylib/src/libraylib.a -lgdi32 -lwinmm`

I have yet to try Linux, I'm hoping it's similarly painless.

**Reply by Peter0x44** [reply]
I tried with Linux, I think unfortunately it's using my mesa provided libEGL.so

```
INFO: GL: OpenGL device information:
INFO:     > Vendor:   AMD
INFO:     > Renderer: AMD Radeon RX Vega (vega10, LLVM 15.0.7, DRM 3.52, 6.3.5-arch1-1)
INFO:     > Version:  OpenGL ES 3.2 Mesa 23.1.1
INFO:     > GLSL:     OpenGL ES GLSL ES 3.20
```
I stepped in with a debugger, and noticed this:
`356             _glfw.egl.handle = _glfwPlatformLoadModule(sonames[i]);`
sonames[0] is libEGL.so.1
maybe I need to edit LD_LIBRARY_PATH or try hardcoding an absolute path here in GLFW. I'll give it some more investigation later

**Reply by grplyler** [reply]
Not a bad strategy... Yeah, Chromium based browsers all use ANGLE. Will probably mention that in my writeup! What Linux Distro you using? I might give it a shot. One of the benefits of building ANGLE from scratch is probably tuning and optimizing. But yeah, this is probably the easier route for most.

**Reply by Peter0x44** [reply]
ah, I fixed it!!!
For some reason, on Linux, by default, GLFW looks for libEGL.so.1 specifically
I had to modify it to:
```patch
diff --git a/src/external/glfw/src/egl_context.c b/src/external/glfw/src/egl_context.c
index bc5f3e60..4700e573 100644
--- a/src/external/glfw/src/egl_context.c
+++ b/src/external/glfw/src/egl_context.c
@@ -343,7 +343,7 @@ GLFWbool _glfwInitEGL(void)
 #elif defined(__OpenBSD__) || defined(__NetBSD__)
         "libEGL.so",
 #else
-        "libEGL.so.1",
+        "libEGL.so",
 #endif
         NULL
     };
```
Then after compiling raylib with GRAPHICS=GRAPHICS_API_OPENGL_ES2, I could use this command:
`gcc -I. ../examples/core/core_basic_window.c libraylib.a /usr/lib/chromium/libEGL.so /usr/lib/chromium/libGLESv2.so  -lm -lpthread -ldl -lrt -lX11`
WARNING: raylib's makefile doesn't rebuild GLFW when you change it, must do make clean

I ran the executable with:
`LD_LIBRARY_PATH=/usr/lib/chromium ./a.out`
and saw the proper output:
```
INFO: GL: OpenGL device information:
INFO:     > Vendor:   Google Inc. (AMD)
INFO:     > Renderer: ANGLE (AMD, AMD Radeon RX Vega (vega10 LLVM 15.0.7), OpenGL 4.6 (Core Profile) Mesa 23.1.1)
INFO:     > Version:  OpenGL ES 3.0.0 (ANGLE 2.1.0 git hash: unknown hash)
INFO:     > GLSL:     OpenGL ES GLSL ES 3.00 (ANGLE 2.1.0 git hash: unknown hash)
```
changing the LD_LIBRARY_PATH is required, because otherwise glfw will still dlopen the mesa provided libEGL.so, despite that I was passing paths to them on the command line... I'm not really sure why, I guess I have more to learn about how shared libs work here.

**Reply by Peter0x44** [reply]
> Will probably mention that in my writeup!

I think it'd be best if all OS were just consolidated to a single page.
I'd like to discuss this further what to put on the wiki page in the raylib discord, if you're up for it.

**Reply by Peter0x44** [reply]
> One of the benefits of building ANGLE from scratch is probably tuning and optimizing.

This is just conjecture, I haven't ran any actual benchmark, but I suspect there would not be such a disparity in performance between them, apple's opengl drivers are exceptionally terrible, I doubt such a difference would be observed on either windows or linux.
But, I would love to be proved wrong.

**Reply by wwderw** [reply]
> changing the LD_LIBRARY_PATH is required, because otherwise glfw will still dlopen the mesa provided libEGL.so, despite that I was passing paths to them on the command line... I'm not really sure why, I guess I have more to learn about how shared libs work here.


Might be able to change that with patchelf, but I would probably put it somewhere to where it's more portable.  I know with Zig build system, can set the rpath in the build file itself.  I'm not too good with make or cmake otherwises.  Having said that, I still have a problem with the binary seeing it.  Still segfaults looking for it, even though ldd is showing the correct path, so I dunno.

**Reply by wwderw** [reply]
This is interesting.  Apparently my issues to linking are due to I had setup for both X11 and Wayland, but when I compiled the library for just X11, everything linked as it should.  I guess Angle must be setup for X11 and not Wayland or both.

**Comment by andymule** [comment]
Eyeing that iOS Metal translation that ANGLE seems to provide like 🕵️‍♀️🕵️🕵️‍♂️

**Comment by paulocoutinhox** [comment]
Nice. It can be merged into raylib and officially add support for macOS and iOS.

**Reply by raysan5** [reply]
Linking with ANGLE is just a build option, no need to change anything on raylib side, afaik. iOS support requires a platform layer that is not implemented, that's the real work to do; the graphics layer is already implemented and probably working thanks to ANGLE.

In any case, contributors are welcome to implement the iOS platform layer, I have no plans to do it and in 10 years nobody has done it.

**Comment by grplyler** [comment]
Writeup on how to do this is out! Also got the source code at https://github.com/grplyler/raylib-articles if we want to collab, @Peter0x44 on adding it to the Raylib wiki.

Medium post: https://medium.com/@grplyler/building-and-linking-googles-angle-with-raylib-on-macos-67b07cd380a3
Dev.to post: https://dev.to/grplyler/building-and-linking-googles-angle-with-raylib-on-macos-11kp

**Reply by grplyler** [reply]
Update: I also added to my posts how to compile ANGLE with _only_ the metal backend. No need for D3D, Vulkan, etc on mac. And when linking and running, its interesting, i get slightly different debug info on the renderer:

```
INFO: Initializing raylib 4.6-dev
INFO: Supported raylib modules:
INFO:     > rcore:..... loaded (mandatory)
INFO:     > rlgl:...... loaded (mandatory)
INFO:     > rshapes:... loaded (optional)
INFO:     > rtextures:. loaded (optional)
INFO:     > rtext:..... loaded (optional)
INFO:     > rmodels:... loaded (optional)
INFO:     > raudio:.... loaded (optional)
INFO: DISPLAY: Trying to enable VSYNC
INFO: DISPLAY: Device initialized successfully
INFO:     > Display size: 1920 x 1080
INFO:     > Screen size:  800 x 450
INFO:     > Render size:  800 x 450
INFO:     > Viewport offsets: 0, 0
INFO: GLAD: OpenGL ES2.0 loaded successfully
INFO: GL: Supported extensions count: 118
INFO: GL: OpenGL device information:
INFO:     > Vendor:   Google Inc. (Apple)
INFO:     > Renderer: ANGLE (Apple, ANGLE Metal Renderer: Apple M1 Pro, Version 12.5 (Build 21G72))
INFO:     > Version:  OpenGL ES 3.0.0 (ANGLE 2.1.21199 git hash: b0e9bbd79fb6)
INFO:     > GLSL:     OpenGL ES GLSL ES 3.00 (ANGLE 2.1.21199 git hash: b0e9bbd79fb6)
INFO: GL: VAO extension detected, VAO functions loaded successfully
INFO: GL: NPOT textures extension detected, full NPOT textures supported
INFO: GL: ETC1 compressed textures supported
INFO: GL: PVRT compressed textures supported
INFO: GL: ASTC compressed textures supported
INFO: TEXTURE: [ID 1] Texture loaded successfully (1x1 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 1] Default texture loaded successfully
INFO: SHADER: [ID 1] Vertex shader compiled successfully
INFO: SHADER: [ID 2] Fragment shader compiled successfully
INFO: SHADER: [ID 3] Program shader loaded successfully
INFO: SHADER: [ID 3] Default shader loaded successfully
INFO: RLGL: Render batch vertex buffers loaded successfully in RAM (CPU)
INFO: RLGL: Render batch vertex buffers loaded successfully in VRAM (GPU)
INFO: RLGL: Default OpenGL state initialized successfully
INFO: TEXTURE: [ID 2] Texture loaded successfully (128x128 | GRAY_ALPHA | 1 mipmaps)
INFO: FONT: Default font loaded successfully (224 glyphs)
^C
```

Also, no matter what i did, I couldn't uncap the framerate with this build! It was limited somewhere in the implementation. Switching to the non-pruned versions of ANGLE and not changing my raylib code at all (not even recompiling) the framerate shot up to 11,000+ FPS. Idk if this is apple or google limiting this kind of GPU resource hogging.

**Reply by raysan5** [reply]
AMAZING article! Thank you very much for writing it! And congratulations again for the accomplishment! It will be really useful for many raylib users! Thanks! 😄

**Reply by raysan5** [reply]
One question, did you check the differences in size/version between ANGLE from Chrome and ANGLE you build?

**Reply by grplyler** [reply]
I couldn't readily find the actual version number I built, but it was from ANGLEs master repo:

```
commit b0e9bbd79fb6c5419ba13627d6fa109c299850fc (HEAD -> main, origin/chromium/5806)
Author: Shahbaz Youssefi <syoussefi@chromium.org>
Date:   Wed May 31 14:23:40 2023 -0400
```

The size of my `libGLESv2.dylib` build with all backends is `11MB`.
The size of my `libGLESv2.dylib` with just the Metal backend is `7.4MB`.
There were two version bundled with Chrome, 
- Version `113.0.5672.126` of `libGLESv2.dylib` is `13MB`
- Version `114.0.5735.90` of `libGLESv2.dylib` is `14MB`

Hope that helps!

**Comment by andrei-zgirvaci** [comment]
@grplyler thanks for the article and research!

I tried the one-line command you showed in the article. It compiles successfully, but I am not sure it's using the ANGLE in the end.
 
This is the one-line command I am using to compile: `c++ -std=c++20 main.cpp -o main $(pkg-config --libs --cflags raylib | string split " ") /Applications/Google\ Chrome.app/Contents/Frameworks/Google\ Chrome\ Framework.framework/Versions/*/Libraries/libEGL.dylib /Applications/Google\ Chrome.app/Contents/Frameworks/Google\ Chrome\ Framework.framework/Versions/*/Libraries/libGLESv2.dylib`

I think the issue is that it doesn't choose the ANGLE library and defaults to openGL. I tried to add `-DCUSTOMIZE_BUILD=ON -DOPENGL_VERSION="ES 2.0"` at the end of the command, it compiles, but nothing changes. Maybe it's only valid for cmake...

Log:
```
INFO: GLAD: OpenGL extensions loaded successfully
INFO: GL: Supported extensions count: 43
INFO: GL: OpenGL device information:
INFO:     > Vendor:   Apple
INFO:     > Renderer: Apple M2
INFO:     > Version:  4.1 Metal - 83.1
INFO:     > GLSL:     4.10
```

I would appreciate any help, thanks!

**Comment by gen2brain** [comment]
@andrei-zgirvaci Those are only cmake valid options. You need to define `-DGRAPHICS_API_OPENGL_ES2`.

**Reply by andrei-zgirvaci** [reply]
Doesn't seems to make any difference in the logs. Not sure it's taking any effect.

**Reply by gen2brain** [reply]
You are using `pkg-config`, which will probably again return `-framework OpenGL` for LDFLAGS, so you are also linking OpenGL. This may not be the exact cause but it seems you are not following the article to the point, you didn't copy the libraries to your project dir, etc.

**Comment by gen2brain** [comment]
I just tested ANGLE on Linux with my Go bindings. I used the `-Wl,-rpath` option to not have to pass `LD_LIBRARY_PATH` every time.
The entire build command is `CGO_LDFLAGS="/opt/google/chrome/libEGL.so /opt/google/chrome/libGLESv2.so -Wl,-rpath=/opt/google/chrome" go build -tags angle`, and then the full path to the library is linked, like this:
```
# ldd ./basic_window | grep GL
	libEGL.so => /opt/google/chrome/libEGL.so (0x00007ff928f8b000)
	libGLESv2.so => /opt/google/chrome/libGLESv2.so (0x00007ff928957000)
```
I also had to symlink `libEGL.so` to `libEGL.so.1`, as @Peter0x44 mentioned, GLFW will try to dlopen it with `.so.1`.


As for performance, it is actually a little better with plain OpenGL, around 2500FPS, with ANGLE it is around 2300, on intel GPU:
```
INFO: GLAD: OpenGL ES2.0 loaded successfully
INFO: GL: Supported extensions count: 138
INFO: GL: OpenGL device information:
INFO:     > Vendor:   Google Inc. (Intel)
INFO:     > Renderer: ANGLE (Intel, Mesa Intel(R) Xe Graphics (TGL GT2), OpenGL 4.6 (Core Profile) Mesa 23.1.0)
INFO:     > Version:  OpenGL ES 3.0.0 (ANGLE 2.1.20795 git hash: b02f159baf67)
INFO:     > GLSL:     OpenGL ES GLSL ES 3.00 (ANGLE 2.1.20795 git hash: b02f159baf67)
```

And every time I close the window there is a segfault, that is on X11:
```
Thread 1 "basic_window" received signal SIGSEGV, Segmentation fault.
0x00007ffff76c8e33 in _XSend () from /usr/lib64/libX11.so.6
(gdb) bt
#0  0x00007ffff76c8e33 in _XSend () from /usr/lib64/libX11.so.6
#1  0x00007ffff76bf201 in XQueryExtension () from /usr/lib64/libX11.so.6
#2  0x00007ffff7352bcb in ?? () from /usr/lib64/libGLX.so.0
#3  0x00007ffff735366b in ?? () from /usr/lib64/libGLX.so.0
#4  0x00007ffff734fe44 in glXDestroyPbuffer () from /usr/lib64/libGLX.so.0
#5  0x00007ffff7defc27 in ?? () from /opt/google/chrome/libGLESv2.so
#6  0x0000000000868770 in ?? ()
#7  0x0000000000768580 in ?? ()
#8  0x00007fffffffd188 in ?? ()
#9  0x00007fffffffd180 in ?? ()
#10 0x00007fffffffd160 in ?? ()
#11 0x00007ffff7c2c37d in ?? () from /opt/google/chrome/libGLESv2.so
#12 0x00007fffffffd1b0 in ?? ()
#13 0x0000000000000000 in ?? ()
(gdb) quit
```

**Comment by BigJk** [comment]
Linking with ANGLE works like a charm!

The only problem I faced was that default alpha blending was somehow broken. Everything was way to light and drop shadows looked more like a white glow. I had to use the following custom blend mode for everything to look normal again. I will just leave this here in case anyone else stubbles over that problem.

```c++
BeginDrawing();
{
  ClearBackground(Color{0, 0, 0, 255});

#ifdef __APPLE__
    rlSetBlendFactorsSeparate(0x0302, 0x0303, 1, 0x0303, 0x8006, 0x8006);
    BeginBlendMode(BLEND_CUSTOM_SEPARATE);
#endif

// ...

}
```

**Reply by grplyler** [reply]
Glad to hear it worked! (In general)
